### PR TITLE
Split the _prepare_request_for_build IIB function

### PR DESCRIPTION
**Pull request by midnightercz**
Original the method contained logic for all add, remove, merge api endpoints.
In this commit, this method was split into two methods with same function
but relevant to build and build_merge_index_image modules

### CODE REVIEW

File: iib/workers/tasks/utils.py [hunk 1]
```diff
@@ -280,3 +279,304 @@ def _get_function_arg_value(arg_name, func, args, kwargs):
...
+    if not arches:
+        raise IIBError('No arches were provided to build the index image')
+    return arches
+
+
+class RequestConfig:
+    """
+    Request config abstract class.
+
+    Attributes:
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/utils.py:370]: Could you please fix the doc string to match the rest of the code base?

File: iib/workers/tasks/build_merge_index_image.py [hunk 2]
```diff
@@ -13,20 +13,86 @@
...
     _opm_index_add,
-    _prepare_request_for_build,
     _push_image,
     _update_index_image_build_state,
     _update_index_image_pull_spec,
 )
 from iib.workers.tasks.celery import app
-from iib.workers.tasks.utils import request_logger, run_cmd, set_registry_token
+from iib.workers.tasks.utils import (
+    request_logger,
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build_merge_index_image.py:22]: Could you please put the imports in an alphabetical order?

File: iib/workers/tasks/build_merge_index_image.py [hunk 3]
```diff
@@ -13,20 +13,86 @@
...
+    All information that was retrieved and/or calculated for the next steps in the build are
+    returned as a dictionary.
+
+    This function was created so that code didn't need to be duplicated for the ``add`` and ``rm``
+    request types.
+
+    :param int request_id: the ID of the IIB build request
+    :param RequestConfg build_request_config: build request configuration
+    :rtype: dict
+    :raises IIBError: if the container image resolution fails or the architectures couldn't be
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build_merge_index_image.py:52]: Could you please have this after `:return:`

File: iib/workers/tasks/build_merge_index_image.py [hunk 4]
```diff
@@ -13,20 +13,86 @@
...
+        detected.
+    :return: a dictionary with the keys: arches, binary_image_resolved, from_index_resolved, and
+        ocp_version.
+    """
+    set_request_state(request_id, 'in_progress', 'Resolving the container images')
+    infos = get_index_image_infos(
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build_merge_index_image.py:58]: Could you please rename `get_index_image_infos` to `get_index_image_info`

Comment (yashvardhannanavati) [reply iib/workers/tasks/build_merge_index_image.py:58]: Also, could we please rename `infos` to something like `index_image_info` ?

File: iib/workers/tasks/utils.py [hunk 5]
```diff
@@ -280,3 +279,311 @@ def _get_function_arg_value(arg_name, func, args, kwargs):
...
+        )
+        result['resolved_distribution_scope'] = (
+            get_image_label(from_index_resolved, 'com.redhat.index.delivery.distribution_scope')
+            or 'prod'
+        )
+        result['resolved_from_index'] = from_index_resolved
+    return result
+
+
+def get_index_image_infos(build_request_config, index_version_map):
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/utils.py:343]: Could you please rename this to `get_index_image_info`

Comment (yashvardhannanavati) [reply iib/workers/tasks/utils.py:343]: Ah, I see there's another method by the same name. Could you please rename this to `get_all_index_image_info` or `get_multiple_index_image_info`

Comment (yashvardhannanavati) [reply iib/workers/tasks/utils.py:343]: Also, could you please add the docstring?

File: iib/workers/tasks/utils.py [hunk 6]
```diff
@@ -280,3 +279,311 @@ def _get_function_arg_value(arg_name, func, args, kwargs):
...
+        infos[index] = get_index_image_info(
+            build_request_config.overwrite_from_index_token,
+            from_index=getattr(build_request_config, index),
+            default_ocp_version=version,
+        )
+    return infos
+
+
+def gather_index_image_arches(build_request_config, index_image_infos):
+    """Gather architectures from build_request_config and provided index image."""
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/utils.py:356]: Docstring here too please

File: iib/workers/tasks/utils.py [hunk 7]
```diff
@@ -280,3 +279,311 @@ def _get_function_arg_value(arg_name, func, args, kwargs):
...
+    arches = set(build_request_config.add_arches or [])
+    for val in index_image_infos.values():
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/utils.py:358]: How about `for info in....`

File: iib/workers/tasks/utils.py [hunk 8]
```diff
@@ -280,3 +279,311 @@ def _get_function_arg_value(arg_name, func, args, kwargs):
...
+        identify the appropriate ``binary_image`` to use.
+    :type binary_image_config: dict
+    """
+
+    _attrs = [
+        "binary_image",
+        "overwrite_from_index_token",
+        "add_arches",
+        "distribution_scope",
+        "source_from_index",
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/utils.py:400]: If this is meant to be generic, shouldn't `source_from_index` and `target_index` not be here.

File: iib/workers/tasks/utils.py [hunk 9]
```diff
@@ -280,3 +279,311 @@ def _get_function_arg_value(arg_name, func, args, kwargs):
...
+    :param bundles: the list of bundles to create the
+        bundle mapping on the request
+    :type bundles: list
+
+    """
+
+    _attrs = RequestConfig._attrs + ["from_index", "add_arches", "bundles"]
+    __slots__ = _attrs
+
+    def get_binary_image(self, index_info, distribution_scope):
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/utils.py:451]: Why not have this in the base class if both classes use the exact same method?

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]

File: iib/workers/tasks/build.py [hunk 10]
```diff
@@ -764,83 +629,33 @@ def _prepare_request_for_build(
...
+    :param RequestConfg build_request_config: build request configuration
     :rtype: dict
     :raises IIBError: if the container image resolution fails or the architectures couldn't be
         detected.
+    :return: a dictionary with the keys: arches, binary_image_resolved, from_index_resolved, and
+        ocp_version.
     """
-    if bundles is None:
+    bundles = build_request_config.bundles
+    if build_request_config.bundles is None:
```
[the remark this answers is not in the document]
Comment (midnightercz) [reply iib/workers/tasks/build.py:641]: yes, it's also possible to write it like that to save few characters

File: iib/workers/tasks/build_merge_index_image.py [hunk 11]
```diff
@@ -13,20 +13,89 @@
...
+    :param RequestConfg build_request_config: build request configuration
+    :rtype: dict
+    :return: a dictionary with the keys: arches, binary_image_resolved, from_index_resolved, and
+        ocp_version.
+
+    :raises IIBError: if the container image resolution fails or the architectures couldn't be
+        detected.
+    """
+    set_request_state(request_id, 'in_progress', 'Resolving the container images')
+    index_image_infos = get_all_index_image_infos(
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build_merge_index_image.py:59]: Could you please rename `index_image_infos ` to `index_image_info`

File: iib/workers/tasks/utils.py [hunk 12]
```diff
@@ -280,3 +279,315 @@ def _get_function_arg_value(arg_name, func, args, kwargs):
...
+        result['resolved_distribution_scope'] = (
+            get_image_label(from_index_resolved, 'com.redhat.index.delivery.distribution_scope')
+            or 'prod'
+        )
+        result['resolved_from_index'] = from_index_resolved
+    return result
+
+
+def get_all_index_image_infos(build_request_config, index_version_map):
+    """Get image infos of all images in version map.
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/utils.py:344]: `Get image infos` => `Get image info`

File: iib/workers/tasks/utils.py [hunk 13]
```diff
@@ -280,3 +279,315 @@ def _get_function_arg_value(arg_name, func, args, kwargs):
...
+        )
+        result['resolved_distribution_scope'] = (
+            get_image_label(from_index_resolved, 'com.redhat.index.delivery.distribution_scope')
+            or 'prod'
+        )
+        result['resolved_from_index'] = from_index_resolved
+    return result
+
+
+def get_all_index_image_infos(build_request_config, index_version_map):
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/utils.py:343]: Could you please rename this to `get_all_index_images_info`

File: iib/workers/tasks/build.py [hunk 14]
```diff
@@ -764,83 +629,37 @@ def _prepare_request_for_build(
...
-        ``None``.
-    :param str source_from_index: the pull specification of the container image containing the index
-        that will be used as a base of the merged index image.
-    :param str target_index: the pull specification of the container image containing the index
-        whose new data will be added to the merged index image.
-    :param dict binary_image_config: the dict of config required to identify the appropriate
-        ``binary_image`` to use.
-    :return: a dictionary with the keys: arches, binary_image_resolved, from_index_resolved, and
-        ocp_version.
+    :param RequestConfg build_request_config: build request configuration
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build.py:633]: `RequestConfg` => `RequestConfig`

File: iib/workers/tasks/build_merge_index_image.py [hunk 15]
```diff
@@ -13,20 +13,89 @@
...
+    Prepare the request for the index image build.
+
+    All information that was retrieved and/or calculated for the next steps in the build are
+    returned as a dictionary.
+
+    This function was created so that code didn't need to be duplicated for the ``add`` and ``rm``
+    request types.
+
+    :param int request_id: the ID of the IIB build request
+    :param RequestConfg build_request_config: build request configuration
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build_merge_index_image.py:50]: `RequestConfg` => `RequestConfig`

File: iib/workers/tasks/utils.py [hunk 16]
```diff
@@ -280,3 +279,315 @@ def _get_function_arg_value(arg_name, func, args, kwargs):
...
+            from_index=getattr(build_request_config, index),
+            default_ocp_version=version,
+        )
+    return infos
+
+
+def gather_index_image_arches(build_request_config, index_image_infos):
+    """Gather architectures from build_request_config and provided index image.
+
+    :param RequestConfg build_request_config: build request configuration
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/utils.py:364]: `RequestConfg` => `RequestConfig`

File: iib/workers/tasks/utils.py [hunk 17]
```diff
@@ -280,3 +279,315 @@ def _get_function_arg_value(arg_name, func, args, kwargs):
...
+            or 'prod'
+        )
+        result['resolved_from_index'] = from_index_resolved
+    return result
+
+
+def get_all_index_image_infos(build_request_config, index_version_map):
+    """Get image infos of all images in version map.
+
+    :param RequestConfg build_request_config: build request configuration
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/utils.py:346]: `RequestConfg` => `RequestConfig`

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]

File: iib/workers/tasks/build_merge_index_image.py (hunk 11 above)
Comment (midnightercz) [reply iib/workers/tasks/build_merge_index_image.py:59]: method is named infos. Wouldn't that be strange to have info?

**Comment by lcarva** [comment]
I'm having a really hard time reviewing this PR. Having better organization of the commits would certainly help.

**Comment by pbortlov** [comment]
@midnightercz  would it be possible to have only `Split the _prepare_request_for_build IIB function` and make other changes in other PRs? Thank you

**Pull request closed without merging**